### PR TITLE
host: Add steps toward stabilising Percy

### PR DESCRIPTION
**Pull request by backspace**
These are some attempts to reduce the frequency of spurious diffs. Common reasons they happen:

[Fonts are not rendered](https://percy.io/Cardstack-1/-cardstack-host/builds/31554282/changed/1736893125?browser=firefox&browser_ids=48%2C49&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=original&width=1280&widths=1280)

screencast 2023-12-07 16-03-51

[Rendering is still in progress](https://percy.io/Cardstack-1/-cardstack-host/builds/31534672/changed/1735964506?browser=firefox&browser_ids=48%2C49&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=1280&widths=1280)

screencast 2023-12-07 16-10-46

“ignored region” no longer covers unstable content

image

Monaco rendering is inconsistent

screencast 2023-12-08 09-31-47

# Mitigation attempts

**Changing to WOFF2 fonts** as “[the best format of the bunch](https://css-tricks.com/understanding-web-fonts-getting/#aa-font-formats) because it offers smaller file sizes and better performance for modern browsers that support it”

**Waiting for fonts** [in a wrapper of `percySnapshot`](https://github.com/cardstack/boxel/pull/883/files#diff-f97ec0793d1764e3c9814900ca4c199835860d063c61c09f8324f5e0de9bc85cR15) just to be sure!

**Waiting for settled state** [in the wrapper](https://github.com/cardstack/boxel/pull/883/files#diff-f97ec0793d1764e3c9814900ca4c199835860d063c61c09f8324f5e0de9bc85cR8) to ensure rendering has completed.

**Removing redundant snapshots** reduces the likelihood of false positives, detailed [here](https://github.com/cardstack/boxel/pull/883#issuecomment-1846063098)

**Adding Percy-specific CSS** to [hide unstable elements](https://github.com/cardstack/boxel/pull/883/files#diff-4f7f310ac8dcb8681c4d7d441895020e2b1fe1d1fd785b7a6b74571a0521f280R5), which is more reliable than using ignored regions.

# What should be snapshotted?

To me, the purpose of snapshots is to detect visual regressions through comparison of a broad selection of UI state across the application, not an exhaustive collection of states. This is informed by the known difficulty of 100% stability in rendering; if we could rely on that, having a snapshot for everything would be fine because it would never produce diff noise.

Looking at the [removed modal snapshots below](#modal-removals) as an example, if we are testing the string contents of fields and their labels, comparing them visually is redundant because they’re just different strings, not different components.

It’s still subjective but when considering adding a new snapshot we can survey what already exists and see if this context differs enough to warrant one.

# Process suggestion

Despite the temptation to just “be done” when submitting a PR for review, we should not casually approve spurious diffs. This just passes on the problems to future us. Instead we should treat these like the flaky tests that they are: open issues so they can be addressed.

If there are issues happening with Percy that seem beyond our control we can also contact their support, they sometimes have helpful suggestions.

### CODE REVIEW

**Comment by backspace** [comment]
An explanation of my editorial decisions on removing snapshots:

<table>
<thead>
<tr><th>removed</th><th rowspan=2></th></tr>
<tr><th>remaining snapshot that suffices, or explanation</th></tr>
</thead>
<tbody>
<tr>
<td>
Integration | card-copy | copy button appears when right and left stacks are index cards and there are selections on right side

Integration | card-copy | copy button appears when right and left stacks are index cards and there are selections on left side
</td>
<td rowspan=2>

screencast 2023-12-07 14-19-11

</td>
</tr>
<tr>
<td>
Integration | card-copy | copy button appears when right and left stacks are index cards and there are mulitple selections on one side
</td>
</tr>
<tr>
<td>
Integration | card-delete | can delete a card from the index card stack item
</td>
<td rowspan=2>

Build #2712 - Cardstack - Percy | Visual testing as a service 2023-12-07 14-21-40

</td>
</tr>
<tr>
<td>The delete button is mostly hidden, this is mostly just exercising the index card</td>
</tr>

<tr>
<td>

Acceptance | code submode | inspector tests | can handle the situation where there is a class name collision with the inherited cards class name

</td>
<td rowspan=2>

screencast 2023-12-07 14-30-46

</td>
</tr>
<tr name='modal-removals'>
<td>

Acceptance | code submode | inspector tests | can inherit from an exported card def declaration

</td>
</tr>

<tr>
<td>

Acceptance | code submode | create-file tests | can create a new card definition in different realm than inherited definition

Acceptance | code submode | create-file tests | can create a new card definition in same realm as inherited definition

Acceptance | code submode | create-file tests | can create a new definition that extends card definition which uses default export

Acceptance | code submode | create-file tests | can reconcile a classname collision with the selected name of extending a card definition which uses a default export

</td>
<td rowspan=2>

screencast 2023-12-07 14-37-46

</td>
</tr>
<tr>
<td>

Acceptance | code submode | create-file tests | can create a new field definition that extends field definition that uses default export

</td>
</tr>

<tr>
<td>

Integration | operator-mode | toggles mode switcher

</td>
<td rowspan=2>

Build #2712 - Cardstack - Percy | Visual testing as a service 2023-12-07 14-45-14

</td>
</tr>
<tr>
<td>

The mode-switcher isn’t visible thanks to being in a wormhole

</td>
</tr>

<tr>
<td>

Integration | operator-mode | displays cards on cards-grid and includes `catalog-entry` instances

</td>
<td rowspan=2>

screencast 2023-12-07 14-47-50

</td>
</tr>
<tr>
<td>

Acceptance | operator mode tests | visiting index card and entering operator mode

</td>
</tr
</tbody>
</table>

File: packages/host/.percy.js [hunk 1]
```diff
@@ -2,5 +2,14 @@ module.exports = {
...
   snapshot: {
     widths: [1280],
+    percyCSS: `
+      .monaco-container[data-test-editor] {
+        display: none;
+      }
+
+      [data-test-definition-info-test], [data-test-last-modified] {
+        visibility: hidden;
+      }
```
Comment (backspace) [inline packages/host/.percy.js:8]: Would something like `.hide-from-percy` be better? Or `[data-test-hide-from-percy]` so it’s not present in production?

Comment (habdelra) [reply packages/host/.percy.js:8]: a test selector would be great!

**Review by habdelra — APPROVED** [review]
looks great! thanks for the great PR description and comments

File: packages/host/.percy.js (hunk 1 above)
Comment (backspace) [reply packages/host/.percy.js:8]: done now, thanks 😀

**Pull request merged**